File: 301-400/383-ransom-note.cpp
// Solution 1 two vectors

bool canConstruct(string ransomNote, string magazine) {
    vector<char> ransom, maga;
    ransom.assign(ransomNote.begin(), ransomNote.end());
    maga.assign(magazine.begin(), magazine.end());

    sort(ransom.begin(), ransom.end());
    sort(maga.begin(), maga.end());

    auto rp = ransom.begin();

    for (auto mb = maga.begin(); mb != maga.end(); ++mb){
        if (rp == ransom.end()) return true;
        if (*rp == *mb) ++rp;
    }
    if (rp == ransom.end()) return true;
    else return false;
}

// Solution 2 hash table

bool canConstruct(string ransomNote, string magazine) {
    int *a = new int[26]();
    for (char c : magazine){
        ++a[c-'a'];
    }
    for (char c : ransomNote){
        if (a[c-'a'] == 0) return false;
        --a[c-'a'];
    }
    return true;
}

File: python version/383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        letter_count = {}
        for i in range(len(magazine)):
            letter_count[magazine[i]] = letter_count.get(magazine[i], 0) + 1

        for i in range(len(ransomNote)):
            letter_count[ransomNote[i]] = letter_count.get(ransomNote[i], 0) - 1
            if letter_count[ransomNote[i]] < 0: return False

        return True